pkg/grpc/client.go

```diff
@@ -144,3 +146,46 @@ func grpcTimeoutInterceptor(timeout time.Duration) grpc.UnaryClientInterceptor {
 		return invoker(ctx, method, req, reply, cc, opts...)
 	}
 }
+
+type grpcClientConnRecord struct {
+	conn *grpc.ClientConn
+	opts *Options
+}
+
+var grpcClientConns = struct {
+	sync.Mutex
+	m map[string]grpcClientConnRecord
+}{
+	m: make(map[string]grpcClientConnRecord),
+}
+
+// GetGRPCClientConn returns a gRPC client connection for the given name. If a connection for that name has already been
+// established the existing connection will be returned. If any options change for that connection, the existing
+// connection will be closed and a new one established.
+func GetGRPCClientConn(name string, opts *Options) (*grpc.ClientConn, error) {
+	grpcClientConns.Lock()
+	defer grpcClientConns.Unlock()
+
+	current, ok := grpcClientConns.m[name]
+	if ok {
+		if cmp.Equal(current.opts, opts) {
+			return current.conn, nil
+		}
+
+		err := current.conn.Close()
+		if err != nil {
+			log.Error().Err(err).Msg("grpc: failed to close existing connection")
```

Comment:
I think we should `Warn` only, as we are going to ignore this anyway.

Reply:
Counterpoint, we "ignore" if there is an error here too, but I definitely think this should be bubbled up as an error.

https://github.com/pomerium/pomerium/pull/1280/files#diff-3ac192716a1df66bbd1008d327af8792R93